internal/store/postgres/rule.go

```diff
@@ -9,80 +9,89 @@ import (
 	"gorm.io/gorm"
 )
 
-type RuleResponse struct {
-	NamespaceUrn string
-	ProviderUrn  string
-	ProviderType string
-	ProviderHost string
-}
-
 // RuleRepository talks to the store to read or insert data
 type RuleRepository struct {
-	*transaction
+	client *Client
 }
 
 // NewRuleRepository returns repository struct
-func NewRuleRepository(db *gorm.DB) *RuleRepository {
-	return &RuleRepository{&transaction{db}}
+func NewRuleRepository(client *Client) *RuleRepository {
+	return &RuleRepository{client}
 }
 
-func (r *RuleRepository) Upsert(ctx context.Context, rl *rule.Rule) error {
+func (r *RuleRepository) UpsertWithTx(ctx context.Context, rl *rule.Rule, postProcessFn func([]rule.Rule) error) error {
```

Comment:
the last refactor using transaction via context was to avoid approach like this (passing a function for post processing). The reason for that was because post-processing should be part of the business logic and should only implemented in the service layer, so repository layer doesn't have to care about anything aside from db communication.

Reply:
That's true about post processing. This is how it is working now. DB layer does not need to know logic of post processing (whether there is a post process or a no-op).

Passing transaction in context would become problematic if the transaction is a long-lived transaction. Context shouldn't be treated as a bag of objects.

In the previous implementation (even before thr refactor that you mentioned), there were some operations that doesn't need transaction. Transaction is expensive so we need to carefully thought about it.